### PR TITLE
feat: Add optional return normalization to JointPPOTrainer for #159 ablation

### LINKED ISSUE #159: P3 plateau: try return normalization + value-coef sweep to unblock PPO learning
## Context

Phase 1 diagnostics (#145 → commit ce841db3, plots and summary in `experiments/p3_specialization/diagnostics/`) confirmed *quantitatively* that the JointPPOTrainer's plateau is dominated by an enormous value-loss term:

| scenario | reward iter 0 → 49 | random baseline | value_term / policy_term at iter 0 |
|---|---|---|---|
| trivial_cooperation | 400.44 → 400.29 | 400.0 | 1.1 × 10⁷ |
| default | 293.46 → 294.58 | 308.0 | 5.0 × 10⁶ |
| chain_reaction | 224.21 → 224.47 | 233.0 | 3.0 × 10⁶ |

The value loss (`vf_coef=0.5 * MSE(V, returns)`) is six to seven orders of magnitude larger than the clipped policy-loss term. The optimizer is fitting the value head and effectively ignoring the policy gradient. Action-distribution entropy also collapses early (~0.66 → ~0.13 nats over 50 iters).

This blocks every quantitative claim from the P3/P1/P4 sweeps: there's no learned policy for the redundancy regularizer to act on, so any signal we measure is a property of the random-init + value-head-fitting trajectory, not of "specialization." See research_notebook/2026-05-14_p3_specialization_results.md for the full writeup.

## Asks

This issue is the **Phase 2 fix** to #145. Phase 1 was diagnosis; this is the targeted remediation. Three orthogonal levers, plus a small ablation that disambiguates which one mattered:

1. **Return normalization in JointPPOTrainer.**
   Track a running mean/std of returns (e.g., across all agents, rolling over the last K rollouts), normalize returns before computing MSE, and re-scale the value prediction at inference. This shrinks the value loss to O(1) and is the cleanest answer to the order-of-magnitude mismatch.

2. **Lower vf_coef.**
   Default `vf_coef=0.5` was inherited from cartpole-scale rewards (|return| ~ 100). With our team rewards in [-400, +400] per step and ~13-night episodes, the natural value-head scale is 10³-10⁴. Sweep `vf_coef ∈ {0.5, 0.05, 0.005, 0.0005}` at λ=0 and pick the value that lets reward actually improve.

3. **Entropy-coef bump (and maybe a schedule).**
   `entropy_coef=0.01` doesn't keep the action distribution open. The entropy term is ~10⁷× smaller than the value term at iter 0, so there is no effective exploration pressure. Try `entropy_coef ∈ {0.01, 0.05, 0.2}`, optionally with a linear decay from a higher starting value.

4. **Ablation.**
   Run all three on `default` at λ_red=0, 5 seeds each, 50 iters × 2048 steps. Report final reward, final entropy, fraction of episodes scoring above the random baseline (308). Pick the smallest-change combination that lands above baseline.

**Constraint.** Run the ablation on `rwalters-sandbox-1`, not locally on the Mac Studio (the active research operating posture: keep the sweep host clean for the next protocol run; iterate algorithms in a sandbox).

**Don't re-launch the P3 sweep yet.** Settle the value/entropy story first. P3 v2 should use whatever Phase 2 lands on.

## Acceptance criteria

- One of the three levers (alone or combined) drives default-scenario final-iter mean reward above 320 (vs random baseline 308) within 50 iters × 2048 steps × 5 seeds.
- The fix is **one toggleable change** in JointPPOTrainer (e.g., `normalize_returns: bool`, or a default `vf_coef` adjustment), with the prior behavior recoverable via a flag.
- A short notebook entry under research_notebook/ that updates the table from #151 with the post-fix numbers.

## Related

- #145 — Phase 1 plateau diagnosis
- #146 — Vacuous CMI conditioning (separate but compounding problem)
- #135 — Rust core reward attribution (relevant once we re-run the sweep on the thrust path)
- #153 / PR #164 — **Levers 2 + 3 plumbing landed (2026-05-14).** `value_coef` and `entropy_coef` are now on `CellConfig` and exposed via CLI flags on `train.py` and `run_sweep.py`. Levers 2 and 3 from this issue can now be swept with no further code changes. **Lever 1 (return normalization) remains unimplemented and is the primary scope of this issue going forward.**

---

### Classification

**Superset of #153** (Category B). #159 was written before #153 and listed three levers; #153 split off as a focused ticket for levers 2 + 3 (value_coef / entropy_coef plumbing) and landed via PR #164. **What remains for #159 is lever 1 (return normalization) plus the ablation that combines all three levers.**

### What #153 / PR #164 already delivered

PR #164 added (verified against `main` 2026-05-14):

- `CellConfig.value_coef: float = 0.5` (`experiments/p3_specialization/train.py:61`)
- `CellConfig.entropy_coef: float = 0.01` (`experiments/p3_specialization/train.py:62`)
- Plumbed through to `JointPPOTrainer(...)` constructor at `train.py:190-191`
- `--value-coef` / `--entropy-coef` CLI flags on both `train.py` (`:261, :271`) and `run_sweep.py` (`:113, :123`)

`JointPPOTrainer.__init__` already accepted these kwargs (`bucket_brigade/training/joint_trainer.py:130-131`); the gap was only in the experiment driver. Defaults are unchanged (so #164 is plumbing-only, no behavior change).

**Net effect for #159**: levers 2 and 3 are now sweepable from the CLI without touching code. A coefficient-only ablation can be launched immediately on sandbox-1.

### What return normalization adds (lever 1)

`grep -n 'normalize\|running_mean\|RunningMeanStd' bucket_brigade/training/joint_trainer.py` confirms: **only advantage normalization exists** (`joint_trainer.py:353`, `advantages[i] = (adv - adv.mean()) / (adv.std() + 1e-8)`). There is no return or reward normalization. `_compute_advantages` at `joint_trainer.py:318-336` returns raw `adv + values` as the regression target, and `F.mse_loss(values, ret_mb)` at `joint_trainer.py:403` regresses the value head against returns whose magnitude is ~10² to ~10³.

Standard PPO return normalization (as in Stable-Baselines3 `VecNormalize`, RLlib's `normalize_returns`, OpenAI baselines):

1. Maintain a `RunningMeanStd` over **returns** (Welford / Chan online algorithm) updated each rollout.
2. Divide returns by `sqrt(var + ε)` before computing `F.mse_loss(values, returns)`.
3. Either (a) train the value head against the normalized target and rescale at inference time, or (b) maintain a running scale that the value head implicitly learns to predict in normalized space and rescale only when computing advantages.

This is **structurally different** from levers 2/3 (which scale the loss term) — normalization shrinks the *regression target itself* to O(1), eliminating the magnitude problem at its source. The two are complementary: returning to `vf_coef=0.5` is appropriate **once** returns are normalized.

### Recommended implementation pointers

- **New flag on `JointPPOTrainer.__init__`**: `normalize_returns: bool = False` (preserves prior behavior when False).
- **Module-level helper**: implement `RunningMeanStd` (Welford) or import from a tiny utility; one global instance per `JointPPOTrainer` shared across agents is fine since their returns are commensurate.
- **Update point**: in `update()` at `bucket_brigade/training/joint_trainer.py:344-354`, after computing `returns[i]`, update the running stats with the concatenation of all agent returns and divide each `returns[i]` by `sqrt(var + 1e-8)`.
- **Inference rescaling**: not strictly required for training correctness (advantages are already computed before normalization is applied; only the value-head MSE target is scaled). If the value head is used elsewhere (it isn't right now — values feed only GAE and the MSE), multiply predictions by `sqrt(var)` at the consumption site.
- **Plumbing**: mirror PR #164's pattern. Add `CellConfig.normalize_returns: bool = False` and `--normalize-returns` CLI flags on `train.py` and `run_sweep.py`.

### Recommended next steps

1. **Implement return normalization** (lever 1) as the primary deliverable for this issue. Builder PR should match #164's plumbing pattern (CellConfig field + CLI flag on train.py and run_sweep.py + JointPPOTrainer kwarg with default False).
2. **Run the ablation on rwalters-sandbox-1** in tmux: `default` scenario, λ_red=0, 5 seeds × 50 iters × 2048 steps. Compare four cells: (a) baseline (no changes), (b) levers 2+3 only (lower `value_coef`, higher `entropy_coef`), (c) lever 1 only (return normalization, default coefs), (d) all three. This disambiguates which mechanism is sufficient.
3. **If lever 1 alone clears baseline+12** (default > 320), recommend committing return normalization as default-on; otherwise propose the smallest combined change in a Phase 3 follow-up.
4. **Update research notebook** with the post-fix table.

## Affected Files

- `bucket_brigade/training/joint_trainer.py` — add `normalize_returns: bool = False` kwarg to `__init__` (after line 131); maintain `RunningMeanStd` state; apply to `returns[i]` in `update()` between current lines 352 and 354. Optionally add a small `RunningMeanStd` helper class above `JointPPOTrainer` or in a sibling module.
- `experiments/p3_specialization/train.py` — add `normalize_returns: bool = False` to `CellConfig` (around line 62, beside the new coef fields from #164); plumb into `JointPPOTrainer(...)` call (around line 190); add `--normalize-returns` CLI flag in `main()` (mirror lines 261/271 from #164).
- `experiments/p3_specialization/run_sweep.py` — add `--normalize-returns` CLI flag and propagate through `run_sweep` signature (mirror lines 43/113/144 from #164).
- `experiments/p3_specialization/diagnostics/phase2_summary.md` — new (or `phase2b_summary.md` if #153's phase2 sweep landed first); table with the four ablation cells (baseline, coefs-only, norm-only, all-three).
- `research_notebook/2026-05-14_p3_specialization_results.md` — append a "Phase 2 / lever 1: return normalization" section.

## Test Plan

- [ ] Manual smoke: `uv run python -m experiments.p3_specialization.train --scenario trivial_cooperation --lambda-red 0.0 --seed 42 --num-iterations 5 --normalize-returns --output-dir /tmp/p3_norm_smoke` — verify `metrics.json` shows `value_loss` reduced by ~5-7 OOM vs Phase 1 numbers in `diagnostics/summary.md`.
- [ ] Manual control: same command without `--normalize-returns` — verify `value_loss` magnitude matches Phase 1 (regression check that the default-off flag preserves prior behavior).
- [ ] Automated tests: add unit test in `tests/test_joint_trainer.py` (create if missing) that constructs `JointPPOTrainer(..., normalize_returns=True)`, runs one `update()`, asserts `value_loss` is O(1).
- [ ] Round-trip test: `CellConfig(scenario='default', lambda_red=0.0, seed=0, normalize_returns=True).normalize_returns == True` and CLI `--normalize-returns` parses to `True`.
- [ ] Edge cases: confirm `normalize_returns=True` with `value_coef=0` still trains (no value gradient, normalization no-op for loss); confirm first-rollout numerical stability when `RunningMeanStd` has only one sample (use `ε` in the denominator).
- [ ] Remote verification: ablation sweep (4 cells × 5 seeds × 50 iters × 2048 steps on default scenario) completes in tmux on sandbox-1 without OOM/crash; diagnostics script runs against new tree.

## Out of Scope

- Replacing PPO with another algorithm.
- Network architecture changes (hidden_size, encoder structure).
- Reward shaping or environment changes.
- Re-running the full P3/P1/P4 sweep — Phase 3 follow-up after this lands.
- Changing default coefficients in `joint_trainer.py` — defer to Phase 3 based on ablation results.

## Dependencies

- [x] #153 / PR #164 — value_coef and entropy_coef plumbing (merged 2026-05-14)

### PR DESCRIPTION
**Pull request by rjwalters**
Closes #159 (implementation + plumbing; 4-cell ablation sweep on sandbox-1 still required for validation).

## Summary

Adds optional return normalization to `JointPPOTrainer` per #159 hypothesis (value loss ~10^6-10^7 larger than policy loss per Phase 1 diagnostics).

- New `bucket_brigade/training/normalizers.py`: Welford/Chan `RunningMeanStd` (online mean+var, scalar or vector)
- `JointPPOTrainer.__init__`: new `normalize_returns: bool = False` flag (default OFF = behavior unchanged)
- When enabled: maintains shared running mean/std across all agents' returns; divides each agent's returns by `sqrt(var + 1e-8)` before the value-loss MSE
- Plumbed through `CellConfig` and `--normalize-returns` CLI flag in both `experiments/p3_specialization/train.py` and `run_sweep.py`

## Test plan

Smoke tests confirmed:
- `RunningMeanStd` produces correct mean/std on synthetic input
- `CellConfig().normalize_returns == False` (default preserved)
- Both `train.py --help` and `run_sweep.py --help` show the new flag

Actual 4-cell ablation (baseline / coefs-only / norm-only / all-three) requires sandbox-1 per CLAUDE.md.

### CODE REVIEW

**Comment by rjwalters** [comment]
Changes Requested - lint-python CI failure

The implementation is solid, but the `lint-python` (ruff format check) and `pre-commit` jobs are failing in CI because `experiments/p3_specialization/run_sweep.py` is not ruff-formatted. The offending line is the new `normalize_returns` kwarg default — it exceeds line length and ruff wants the subscript broken across lines.

**Required diff** (verified locally with `uv run ruff format`):

```python
-    normalize_returns: bool = CellConfig.__dataclass_fields__["normalize_returns"].default,
+    normalize_returns: bool = CellConfig.__dataclass_fields__[
+        "normalize_returns"
+    ].default,
```

Run `uv run ruff format experiments/p3_specialization/run_sweep.py` to auto-fix, then push. After this, `lint-python` should go green. The `pre-commit` job also fails on a long list of pre-existing trailing-whitespace / end-of-file issues in `experiments/scenarios/**/*.json`, `.loom/roles/auditor.md`, `.claude/commands/loom/hermit.md`, and `research_notebook/2025-11-08_phase_2_5_analysis_plan.md` — these are not introduced by this PR. Out of scope here.

---

## Review of the PR itself

The code review came out **clean** otherwise:

### RunningMeanStd correctness — verified
Compared against numpy ground truth on a 20-batch stream of `N(5,2)` (mean diff 3.85e-07, var diff 1.67e-06); single 2048-sample batch at PPO-return scale `N(200,100)` (mean diff 9.69e-06, var diff 1.42e-03); and vector shape `(3,)` over 100 samples (max diff 4.98e-08 / 2.92e-07). The Chan parallel-combine in `_update_from_moments` is the correct OpenAI-baselines / SB3 `VecNormalize` formulation: `new_mean = old_mean + delta * b/(a+b)`, `m2 = var_a*a + var_b*b + delta^2 * a*b/(a+b)`, `new_var = m2/(a+b)`. Count accumulates correctly. `epsilon=1e-4` initial pseudo-count is the standard SB3 default.

### Defaults preserved — verified
Loaded `experiments.p3_specialization.train.CellConfig` from the worktree (forcing `sys.path` to the worktree source); `CellConfig(scenario='default', lambda_red=0.0, seed=0).normalize_returns == False`. `JointPPOTrainer.__init__` has `normalize_returns: bool = False` and `self._return_rms = None` on that path, so the cold path has zero overhead.

### Wiring — verified by diff
- Returns from all agents are concatenated to a single `numpy` array before `update()` (correct — they live on the same scale).
- Scale is `sqrt(var + 1e-8)`, applied per-agent divisively.
- Advantages are NOT rescaled — comment on lines 379-381 calls this out explicitly. Correct: advantages are already standardized into the policy loss; only the value-loss MSE target needs shrinking.
- `CellConfig.normalize_returns` plumbed through to `JointPPOTrainer(..., normalize_returns=cfg.normalize_returns)`; `--normalize-returns` flag on both `train.py` and `run_sweep.py` mirrors PR #164's pattern exactly.

### Scope
4 files, 141 insertions / 1 deletion, plumbing-only. The 4-cell ablation (baseline / coefs-only / norm-only / all-three) is correctly deferred to sandbox-1 per CLAUDE.md compute guidelines.

Push the ruff-format fix and I'll re-evaluate immediately.

**Comment by rjwalters** [comment]
Applied ruff format per Judge feedback. Re-requesting review.

**Comment by rjwalters** [comment]
Format fix applied cleanly. LGTM! Approved.
